src/bootstrap/format.rs

```diff
@@ -37,6 +37,103 @@ fn rustfmt(src: &Path, rustfmt: &Path, paths: &[PathBuf], check: bool) -> impl F
     }
 }
 
+fn verify_rustfmt_version(build: &Builder<'_>) -> bool {
+    let stamp_file = {
+        let mut s = build.out.clone();
+        s.push("rustfmt.stamp");
+        s
+    };
+
+    let mut cmd = Command::new(match build.initial_rustfmt() {
+        Some(p) => p,
+        None => return false,
+    });
+    cmd.arg("--version");
+    let output = match cmd.output() {
+        Ok(status) => status,
+        Err(_) => return false,
+    };
+    if !output.status.success() {
+        return false;
+    }
+    let version = String::from_utf8(output.stdout).unwrap();
+    !program_out_of_date(&stamp_file, &version)
+}
+
+fn update_rustfmt_version(build: &Builder<'_>) {
+    let stamp_file = {
+        let mut s = build.out.clone();
+        s.push("rustfmt.stamp");
+        s
+    };
+
+    let mut cmd = Command::new(match build.initial_rustfmt() {
+        Some(p) => p,
+        None => return,
+    });
+    cmd.arg("--version");
+    let output = match cmd.output() {
+        Ok(status) => status,
+        Err(_) => return,
+    };
+    if !output.status.success() {
+        return;
+    }
+    let version = String::from_utf8(output.stdout).unwrap();
+
+    t!(std::fs::write(stamp_file, version))
+}
+
+fn get_modified_files(build: &Builder<'_>) -> Option<Vec<String>> {
+    let Ok(remote) = get_rust_lang_rust_remote() else {return None;};
+    if !verify_rustfmt_version(build) {
+        return None;
+    }
+    Some(
+        output(
+            build
+                .config
+                .git()
+                .arg("diff-index")
+                .arg("--name-only")
+                .arg("--merge-base")
+                .arg(&format!("{remote}/master")),
+        )
+        .lines()
+        .map(|s| s.trim().to_owned())
+        .collect(),
+    )
+}
+
+/// Finds the remote for rust-lang/rust.
+/// For example for these remotes it will return `upstream`.
+/// ```text
+/// origin  https://github.com/Nilstrieb/rust.git (fetch)
+/// origin  https://github.com/Nilstrieb/rust.git (push)
+/// upstream        https://github.com/rust-lang/rust (fetch)
+/// upstream        https://github.com/rust-lang/rust (push)
+/// ```
+fn get_rust_lang_rust_remote() -> Result<String, String> {
```

Comment:
@jyn514 - do you know if we already have this code somewhere? Presumably we do, since we need a similar check for LLVM downloads etc, right?

Reply:
I got this code from #105058

Reply:
Yes, LLVM just hard-codes `origin` currently. In practice both will work, `origin` is just more expensive if it's outdated.